source/NRF52I2C.cpp

```diff
@@ -132,12 +132,22 @@ int NRF52I2C::waitForStop(int evt)
         // If we started a zero length transmission task (typically a bus probe), then timeout
         // after 1ms. The hardware has no way to relay succesful completion, but may return
         // an error condition.
-        if (p_twim->EVENTS_TXSTARTED && p_twim->TXD.MAXCNT == 0 && locked++ == 100)
+        locked++;
+        if (p_twim->EVENTS_TXSTARTED && p_twim->TXD.MAXCNT == 0 && locked >= 100)
         {
             res = DEVICE_OK;
             break;
         }
 
+        // Test for condition where the SHORTS configuration appears to not trigger TASKS as expected.
+        // Could be an undocumented silicon errata.
+        // Appears to only occur under higher levels of background interrupt load.
+        if (locked >= 100 && p_twim->EVENTS_LASTTX && (p_twim->SHORTS & NRF_TWIM_SHORT_LASTTX_SUSPEND_MASK) && !p_twim->EVENTS_SUSPENDED)
+            p_twim->TASKS_SUSPEND = 1;
```

Comment:
Thanks James. I was a bit worried that given the issue is related to the TASK not triggering, spinning on it might be a bad idea... I was relying on the delay a little further down combined with the fact we'd retry again a bit later if it fails.

Perhaps we should reset the `locked` counter though, so it gets a long time to complete, and if it doesn't we try again a millisecond later? Might be a safer option?

Reply:
Ha good point! Yes a 1 ms retry makes sense to me